grafana-plugin/src/utils/authorization/index.ts

```diff
@@ -0,0 +1,151 @@
+import { OrgRole } from '@grafana/data';
+import { contextSrv } from 'grafana/app/core/core';
+
+const ONCALL_PERMISSION_PREFIX = 'grafana-oncall-app';
+
+export type UserAction = {
+  permission: string;
+  fallbackMinimumRoleRequired: OrgRole;
+};
+
+export enum Resource {
+  ALERT_GROUPS = 'alert-groups',
+  INTEGRATIONS = 'integrations',
+  ESCALATION_CHAINS = 'escalation-chains',
+  SCHEDULES = 'schedules',
+  CHATOPS = 'chatops',
+  OUTGOING_WEBHOOKS = 'outgoing-webhooks',
+  MAINTENANCE = 'maintenance',
+  API_KEYS = 'api-keys',
+  NOTIFICATIONS = 'notifications',
+
+  NOTIFICATION_SETTINGS = 'notification-settings',
+  USER_SETTINGS = 'user-settings',
+  OTHER_SETTINGS = 'other-settings',
+
+  TEAMS = 'teams',
+  PLUGINS = 'plugins',
+}
+
+export enum Action {
+  READ = 'read',
+  WRITE = 'write',
+  ADMIN = 'admin',
+  TEST = 'test',
+  EXPORT = 'export',
+  UPDATE_SETTINGS = 'update-settings',
+  INSTALL = 'install',
+}
+
+type Actions =
+  | 'AlertGroupsRead'
+  | 'AlertGroupsWrite'
+  | 'IntegrationsRead'
+  | 'IntegrationsWrite'
+  | 'IntegrationsTest'
+  | 'EscalationChainsRead'
+  | 'EscalationChainsWrite'
+  | 'SchedulesRead'
+  | 'SchedulesWrite'
+  | 'SchedulesExport'
+  | 'ChatOpsRead'
+  | 'ChatOpsWrite'
+  | 'ChatOpsUpdateSettings'
+  | 'OutgoingWebhooksRead'
+  | 'OutgoingWebhooksWrite'
+  | 'MaintenanceRead'
+  | 'MaintenanceWrite'
+  | 'APIKeysRead'
+  | 'APIKeysWrite'
+  | 'NotificationsRead'
+  | 'NotificationSettingsRead'
+  | 'NotificationSettingsWrite'
+  | 'UserSettingsRead'
+  | 'UserSettingsWrite'
+  | 'UserSettingsAdmin'
+  | 'OtherSettingsRead'
+  | 'OtherSettingsWrite'
+  | 'TeamsWrite'
+  | 'PluginsInstall';
+
+const roleMapping: Record<OrgRole, number> = {
+  [OrgRole.Admin]: 0,
+  [OrgRole.Editor]: 1,
+  [OrgRole.Viewer]: 2,
+};
+
+/**
+ * The logic here is:
+ * - an Admin should be able to do everything (including whatever an Editor and Viewer can do)
+ * - an Editor should be able to do things Editors and Viewers can do
+ * - a Viewer is only allowed to do things Viewers can do
+ */
+export const userHasMinimumRequiredRole = (minimumRoleRequired: OrgRole): boolean =>
+  roleMapping[contextSrv.user.orgRole] <= roleMapping[minimumRoleRequired];
+
+/**
+ * See here for more info on the hasAccess method
+ * https://github.com/grafana/grafana/blob/main/public/app/core/services/context_srv.ts#L165-L170
+ *
+ * As a fallback (second argument), for cases where RBAC is not enabled for a grafana instance, rely on basic roles
+ */
+export const isUserActionAllowed = ({ permission, fallbackMinimumRoleRequired }: UserAction): boolean =>
+  contextSrv.hasAccess(permission, userHasMinimumRequiredRole(fallbackMinimumRoleRequired));
```

Comment:
As of now, we could check for the accessControlOnCall feature toggle. 

Once RBAC for OnCall seem to be ready for production and we remove the feature toggle, I'm assuming you will have a version enforcement done upon OnCall installation/update, preventing someone from running a RBAC OnCall on a version of Grafana that doesn't support it.

Wdyt?

Reply:
well this was a conversation @Matvey-Kuk and I were having the other day. I like your idea of just forcing this version of OnCall to only be compatible with >=9.3.x. @Matvey-Kuk's point was that by doing this we may dissuade people to upgrade to the latest version of OnCall if they are running an older version of Grafana and also need to update their grafana.

Nothing changes on the backend in terms of compatibility, it works when RBAC is supported/not supported (by sending a `HEAD` request to your endpoint and seeing if it gets back a 404 or not). This is really the only spot on the frontend where things would be incompatible.

I'm going to do this:
```tsx
export const isUserActionAllowed = ({ permission, fallbackMinimumRoleRequired }: UserAction): boolean => {
  if (config.featureToggles.accessControlOnCall) {
    return !!contextSrv.user.permissions?.[permission];
  }
  return userHasMinimumRequiredRole(fallbackMinimumRoleRequired);
};
```

I'm _fairly certain_ this would allow us to not have to change this in the `plugin.json`:
```json
"dependencies": {
    "grafanaDependency": ">=8.3.2",
    "grafanaVersion": "8.3",
    "plugins": []
  }
```

Reply:
My only worry is that `config.featureToggles.accessControlOnCall` wouldn't exist on `<= 9.3`, isn't it going to prevent frontend to run?

Reply:
I'm confused. I'm using `@grafana/runtime@9.3.0-beta1` which _should_ have this property set, no?

Reply:
The feature toggle is set by the backend starting from 9.3.0, but I don't actually know what happens if you run OnCall on a version (< 9.3) of Grafana that does not have this feature toggle.

Reply:
I'll do some local testing with various versions of grafana to see how this behaves 👍

Reply:
tested locally using `grafana/grafana:latest` image. This works as expected, user's can use this new version of OnCall without having to upgrade their grafana:
```ts
import { config } from '@grafana/runtime';
import { contextSrv } from 'grafana/app/core/core';

...

export const isUserActionAllowed = ({ permission, fallbackMinimumRoleRequired }: UserAction): boolean => {
  if (config.featureToggles.accessControlOnCall) {
    return !!contextSrv.user.permissions?.[permission];
  }
  return userHasMinimumRequiredRole(fallbackMinimumRoleRequired);
};
```